### ISSUE TITLE
I can't verify requests for Discord interactions URL properly

### ISSUE THREAD

**Issue by preslavmihaylov** [bwmarrin/discordgo#1526]
I'm trying to write a golang handler, which reacts to discord interactions. I've set it up as the interactions URL in my discord developer portal.

This is the code I've written:
```
var secrets struct {
	DiscordPublicKey string
}

type UnknownDiscordInteraction struct {
	InteractionType discordgo.InteractionType `json:"type"`
}

func DiscordWebhook(w http.ResponseWriter, r *http.Request) {
	body, err := io.ReadAll(r.Body)
	if err != nil {
		http.Error(w, "Error reading request body", http.StatusInternalServerError)
		return
	}

	publicKeyBytes, err := hex.DecodeString(secrets.DiscordPublicKey)
	if err != nil {
		http.Error(w, "Error decoding hex string", http.StatusInternalServerError)
		return
	}

	if !discordgo.VerifyInteraction(r, publicKeyBytes) {
		http.Error(w, "invalid request signature", http.StatusUnauthorized)
		return
	}

	var interaction UnknownDiscordInteraction
	if err := json.Unmarshal(body, &interaction); err != nil {
		http.Error(w, "Error unmarshalling request body", http.StatusInternalServerError)
		return
	}

	if interaction.InteractionType == discordgo.InteractionPing {
		pongResp, err := json.Marshal(discordgo.InteractionResponse{
			Type: discordgo.InteractionResponsePong,
		})
		if err != nil {
			http.Error(w, "Error marshalling response", http.StatusInternalServerError)
			return
		}

		// set content-type header
		w.Header().Set("Content-Type", "application/json")
		w.Write(pongResp)
		w.WriteHeader(http.StatusOK)
		return
	}

	w.WriteHeader(http.StatusOK)
}
```

I can't save the endpoint in the developer dashboard, because any challenge request that discord sends I fail to verify, it returns that the request is invalid.

I am passing the `DiscordPublicKey` as a string, copied from my discord developer dashboard. Anyone has a clue what I'm doing wrong?

**Comment by piesuke** [comment]
@preslavmihaylov 
I'm facing the same problem.

Did you find a solution?

**Comment by preslavmihaylov** [comment]
No, I ended up writing a proxy in javascript which forwards discord requests to my go service to workaround this

**Comment by Mihitoko** [comment]
I know its late but for the people that see this in the future.
The problem is that you read the body before verifying the interaction.

The request body is a stream once you read all its contents its empty.
Since you read the request body before calling VerifyInteraction you effectively passing an empty request body to the verify function.
Which obviously makes the signatures not match.

VerifyInteraction copies the content it read back into the Request struct, so just read the body after verification is done and you should be fine.

**Comment by ajokunu** [comment]
This saved me after hours of debugging

> I know its late but for the people that see this in the future. The problem is that you read the body before verifying the interaction.
> 
> The request body is a stream once you read all its contents its empty. Since you read the request body before calling VerifyInteraction you effectively passing an empty request body to the verify function. Which obviously makes the signatures not match.
> 
> VerifyInteraction copies the content it read back into the Request struct, so just read the body after verification is done and you should be fine.
